python/variable_init.cpp

```diff
@@ -0,0 +1,456 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Jan-Lukas Wynen
+
+#include <sstream>
+
+#include "pybind11.h"
+
+#include "scipp/core/dtype.h"
+#include "scipp/core/eigen.h"
+#include "scipp/core/tag_util.h"
+#include "scipp/dataset/dataset.h"
+#include "scipp/units/string.h"
+#include "scipp/variable/variable.h"
+
+#include "dtype.h"
+#include "format.h"
+#include "numpy.h"
+#include "py_object.h"
+#include "unit.h"
+
+using namespace scipp;
+using namespace scipp::variable;
+
+namespace py = pybind11;
+
+namespace {
+std::tuple<DType, units::Unit>
+cast_dtype_and_unit(const py::object &dtype,
+                    const std::optional<units::Unit> unit) {
+  const auto scipp_dtype = ::scipp_dtype(dtype);
+  if (scipp_dtype == core::dtype<core::time_point>) {
+    units::Unit deduced_unit = parse_datetime_dtype(dtype);
+    if (unit.has_value()) {
+      if (deduced_unit != units::one && *unit != deduced_unit) {
+        throw std::invalid_argument(format(
+            "The unit encoded in the dtype (", deduced_unit,
+            ") conflicts with the explicitly specified unit (", *unit, ")."));
+      } else {
+        deduced_unit = *unit;
+      }
+    }
+    return std::tuple{scipp_dtype, deduced_unit};
+  } else {
+    return std::tuple{scipp_dtype, unit.value_or(units::one)};
+  }
+}
+
+Dimensions build_dimensions(const py::object &labels, const py::object &shape) {
+  if (labels.is_none()) {
+    if (!shape.is_none()) {
+      throw std::invalid_argument(
+          "Cannot determine dimensions, got a shape but no dimension labels.");
+    }
+    return Dimensions{};
+  } else {
+    if (shape.is_none()) {
+      throw std::invalid_argument(
+          "Cannot determine dimensions, got dimension labels but no shape.");
+    }
+    return Dimensions(labels.cast<std::vector<Dim>>(),
+                      shape.cast<std::vector<scipp::index>>());
+  }
+}
+
+void ensure_consistent_ndim(const std::vector<Dim> &dim_labels,
+                            const py::object &shape,
+                            const std::optional<py::array> &values,
```

Comment:
I think `shape` and `values` were always meant as "mutally exclusive" arguments. Since that is odd, we introduced `sc.zeros` and `sc.arrays` to be used in most user-facing code. Would it be simpler to just throw if both shape and values are given? Is there an advantage to supporting both at the same time?

Reply:
The advantage is that is simplifies generic cases:
```.py
values = get_values()  # could be None
shape = get_shape()
...
var = sc.Variable(shape=shape, values=values, ...)
```

I don't know if it (could) appear in deserialisation but part of my goal of the refactor was to be as generic with the arguments are possible.

Reply:
Seems unlikely, and also yields unclear and error prone code, so I would vote for not supporting this.